File: app/src/main/java/com/aaron/realgltextureview/GLSurfaceView.java
package com.aaron.realgltextureview;

import android.content.Context;
import android.util.AttributeSet;
import android.util.Log;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

import java.lang.ref.WeakReference;

import javax.microedition.khronos.opengles.GL;


public class GLSurfaceView extends SurfaceView implements SurfaceHolder.Callback2, IGLView {
    private final static String TAG = "GLSurfaceView";

    public GLSurfaceView(Context context) {
        super(context);
        init();
    }

    public GLSurfaceView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    @Override
    protected void finalize() throws Throwable {
        try {
            if (mGLThread != null) {
                mGLThread.requestExitAndWait();
            }
        } finally {
            super.finalize();
        }
    }

    private void init() {
        SurfaceHolder holder = getHolder();
        holder.addCallback(this);

    }

    public void setGLWrapper(GLWrapper glWrapper) {
        mGLWrapper = glWrapper;
    }

    public void setDebugFlags(int debugFlags) {
        mDebugFlags = debugFlags;
    }

    @Override
    public EGLConfigChooser getEGLConfigChooser() {
        return mEGLConfigChooser;
    }

    @Override
    public int getEGLContextClientVersion() {
        return 0;
    }

    @Override
    public EGLContextFactory getEGLContextFactory() {
        return null;
    }

    @Override
    public EGLWindowSurfaceFactory getEGLWindowSurfaceFactory() {
        return null;
    }

    @Override
    public Object getSurfaceObject() {
        return null;
    }

    @Override
    public GLWrapper getGLWrapper() {
        return null;
    }

    public int getDebugFlags() {
        return mDebugFlags;
    }

    @Override
    public boolean isPreserveEGLContextOnPause() {
        return false;
    }

    @Override
    public Renderer getRenderer() {
        return null;
    }

    public void setPreserveEGLContextOnPause(boolean preserveOnPause) {
        mPreserveEGLContextOnPause = preserveOnPause;
    }

    public boolean getPreserveEGLContextOnPause() {
        return mPreserveEGLContextOnPause;
    }

    public void setRenderer(Renderer renderer) {
        checkRenderThreadState();
        if (mEGLConfigChooser == null) {
            mEGLConfigChooser = new SimpleEGLConfigChooser(true, mEGLContextClientVersion);
        }
        if (mEGLContextFactory == null) {
            mEGLContextFactory = new DefaultContextFactory();
        }
        if (mEGLWindowSurfaceFactory == null) {
            mEGLWindowSurfaceFactory = new DefaultWindowSurfaceFactory();
        }
        mRenderer = renderer;
        mGLThread = new GLThread(mThisWeakRef);
        mGLThread.start();
    }

    public void setEGLContextFactory(EGLContextFactory factory) {
        checkRenderThreadState();
        mEGLContextFactory = factory;
    }

    public void setEGLWindowSurfaceFactory(EGLWindowSurfaceFactory factory) {
        checkRenderThreadState();
        mEGLWindowSurfaceFactory = factory;
    }

    public void setEGLConfigChooser(EGLConfigChooser configChooser) {
        checkRenderThreadState();
        mEGLConfigChooser = configChooser;
    }

    public void setEGLConfigChooser(boolean needDepth) {
        setEGLConfigChooser(new SimpleEGLConfigChooser(needDepth, mEGLContextClientVersion));
    }

    public void setEGLConfigChooser(int redSize, int greenSize, int blueSize,
                                    int alphaSize, int depthSize, int stencilSize) {
        setEGLConfigChooser(new ComponentSizeChooser(redSize, greenSize,
                blueSize, alphaSize, depthSize, stencilSize, mEGLContextClientVersion));
    }

    public void setEGLContextClientVersion(int version) {
        checkRenderThreadState();
        mEGLContextClientVersion = version;
    }

    public void setRenderMode(int renderMode) {
        mGLThread.setRenderMode(renderMode);
    }

    public int getRenderMode() {
        return mGLThread.getRenderMode();
    }

    public void requestRender() {
        mGLThread.requestRender();
    }

    public void surfaceCreated(SurfaceHolder holder) {
        mGLThread.surfaceCreated();
    }

    public void surfaceDestroyed(SurfaceHolder holder) {
        // Surface will be destroyed when we return
        mGLThread.surfaceDestroyed();
    }

    public void surfaceChanged(SurfaceHolder holder, int format, int w, int h) {
        mGLThread.onWindowResize(w, h);
    }

    @Override
    public void surfaceRedrawNeededAsync(SurfaceHolder holder, Runnable finishDrawing) {
        if (mGLThread != null) {
            mGLThread.requestRenderAndNotify(finishDrawing);
        }
    }

    @Deprecated
    @Override
    public void surfaceRedrawNeeded(SurfaceHolder holder) {
        // Since we are part of the framework we know only surfaceRedrawNeededAsync
        // will be called.
    }

    public void onPause() {
        mGLThread.onPause();
    }

    public void onResume() {
        mGLThread.onResume();
    }

    public void queueEvent(Runnable r) {
        mGLThread.queueEvent(r);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (GLConstant.LOG_ATTACH_DETACH) {
            Log.d(TAG, "onAttachedToWindow reattach =" + mDetached);
        }
        if (mDetached && (mRenderer != null)) {
            int renderMode = GLConstant.RENDERMODE_CONTINUOUSLY;
            if (mGLThread != null) {
                renderMode = mGLThread.getRenderMode();
            }
            mGLThread = new GLThread(mThisWeakRef);
            if (renderMode != GLConstant.RENDERMODE_CONTINUOUSLY) {
                mGLThread.setRenderMode(renderMode);
            }
            mGLThread.start();
        }
        mDetached = false;
    }

    @Override
    protected void onDetachedFromWindow() {
        if (GLConstant.LOG_ATTACH_DETACH) {
            Log.d(TAG, "onDetachedFromWindow");
        }
        if (mGLThread != null) {
            mGLThread.requestExitAndWait();
        }
        mDetached = true;
        super.onDetachedFromWindow();
    }

    private void checkRenderThreadState() {
        if (mGLThread != null) {
            throw new IllegalStateException(
                    "setRenderer has already been called for this instance.");
        }
    }

    private final WeakReference<GLSurfaceView> mThisWeakRef = new WeakReference<GLSurfaceView>(this);
    private GLThread mGLThread;
    private Renderer mRenderer;
    private boolean mDetached;
    private EGLConfigChooser mEGLConfigChooser;
    private EGLContextFactory mEGLContextFactory;
    private EGLWindowSurfaceFactory mEGLWindowSurfaceFactory;
    private GLWrapper mGLWrapper;
    private int mDebugFlags;
    private int mEGLContextClientVersion;
    private boolean mPreserveEGLContextOnPause;
}

File: iglview/src/main/java/com/aaronlee/iglview/GLSurfaceView.kt
package com.aaronlee.iglview

import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.view.SurfaceHolder
import android.view.SurfaceView
import java.lang.ref.WeakReference

class GLSurfaceView : SurfaceView, SurfaceHolder.Callback2, IGLView {
    private val mThisWeakRef: WeakReference<out IGLView?> = WeakReference(this)
    private var mGLThread: GLThread? = null
    private var mRenderer: Renderer? = null
    private var mDetached = false
    private var mEGLConfigChooser: EGLConfigChooser? = null
    private var mEGLContextFactory: EGLContextFactory? = null
    private var mEGLWindowSurfaceFactory: EGLWindowSurfaceFactory? = null
    override var debugFlags = 0
    private var mEGLContextClientVersion = 0
    override var isPreserveEGLContextOnPause = false

    constructor(context: Context) : super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init()
    }

    @Throws(Throwable::class)
    protected fun finalize() {
        try {
            mGLThread?.requestExitAndWait()
        } finally {
//            super.finalize()
        }
    }

    private fun init() {
        val holder = holder
        holder.addCallback(this)
    }

    override var eglConfigChooser: EGLConfigChooser?
        get() = mEGLConfigChooser
        set(configChooser) {
            checkRenderThreadState()
            mEGLConfigChooser = configChooser
        }
    override var eglContextClientVersion: Int
        get() = mEGLContextClientVersion
        set(version) {
            checkRenderThreadState()
            mEGLContextClientVersion = version
        }
    override var eglContextFactory: EGLContextFactory?
        get() = mEGLContextFactory
        set(factory) {
            checkRenderThreadState()
            mEGLContextFactory = factory
        }
    override var eglWindowSurfaceFactory: EGLWindowSurfaceFactory?
        get() = mEGLWindowSurfaceFactory
        set(factory) {
            checkRenderThreadState()
            mEGLWindowSurfaceFactory = factory
        }
    override val surfaceObject: Any?
        get() = holder
    override var renderer: Renderer?
        get() = mRenderer
        set(renderer) {
            checkRenderThreadState()
            if (mEGLConfigChooser == null) {
                mEGLConfigChooser = SimpleEGLConfigChooser(true, mEGLContextClientVersion)
            }
            if (mEGLContextFactory == null) {
                mEGLContextFactory = DefaultContextFactory()
            }
            if (mEGLWindowSurfaceFactory == null) {
                mEGLWindowSurfaceFactory = DefaultWindowSurfaceFactory()
            }
            mRenderer = renderer
            mGLThread = GLThread(mThisWeakRef)
            mGLThread!!.start()
        }

    override fun setEGLConfigChooser(needDepth: Boolean) {
        eglConfigChooser = SimpleEGLConfigChooser(needDepth, mEGLContextClientVersion)
    }

    override fun setEGLConfigChooser(
        redSize: Int, greenSize: Int, blueSize: Int,
        alphaSize: Int, depthSize: Int, stencilSize: Int
    ) {
        eglConfigChooser = ComponentSizeChooser(
            redSize, greenSize,
            blueSize, alphaSize, depthSize, stencilSize, mEGLContextClientVersion
        )
    }

    override var renderMode: Int
        get() = mGLThread?.renderMode?:GLConstant.RENDERMODE_WHEN_DIRTY
        set(renderMode) {
            mGLThread?.renderMode = renderMode
        }

    override fun requestRender() {
        mGLThread?.requestRender()
    }

    override fun surfaceCreated(holder: SurfaceHolder) {
        mGLThread?.surfaceCreated()
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
        // Surface will be destroyed when we return
        mGLThread?.surfaceDestroyed()
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, w: Int, h: Int) {
        mGLThread?.onWindowResize(w, h)
    }

    override fun surfaceRedrawNeededAsync(holder: SurfaceHolder, finishDrawing: Runnable) {
        mGLThread?.requestRenderAndNotify(finishDrawing)
    }

    @Deprecated("")
    override fun surfaceRedrawNeeded(holder: SurfaceHolder) {
        // Since we are part of the framework we know only surfaceRedrawNeededAsync
        // will be called.
    }

    override fun onPause() {
        mGLThread?.onPause()
    }

    override fun onResume() {
        mGLThread?.onResume()
    }

    override fun queueEvent(r: Runnable?) {
        mGLThread?.queueEvent(r)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (GLConstant.LOG_ATTACH_DETACH) {
            Log.d(TAG, "onAttachedToWindow reattach =$mDetached")
        }
        if (mDetached && mRenderer != null) {
            val renderMode = mGLThread?.renderMode?:GLConstant.RENDERMODE_CONTINUOUSLY
            mGLThread = GLThread(mThisWeakRef)
            if (renderMode != GLConstant.RENDERMODE_CONTINUOUSLY) {
                mGLThread?.renderMode = renderMode
            }
            mGLThread?.start()
        }
        mDetached = false
    }

    override fun onDetachedFromWindow() {
        if (GLConstant.LOG_ATTACH_DETACH) {
            Log.d(TAG, "onDetachedFromWindow")
        }
        mGLThread?.requestExitAndWait()
        mDetached = true
        super.onDetachedFromWindow()
    }

    private fun checkRenderThreadState() {
        check(mGLThread == null) { "setRenderer has already been called for this instance." }
    }

    companion object {
        private const val TAG = "GLSurfaceView"
    }
}